### ISSUE TITLE
[Feature Request] Approximate Speculative Decoding (ASD): training-free, budget-bounded acceptance for greedy assisted decoding

### ISSUE THREAD

**Issue by Kissmetothemoon** [huggingface/transformers#48636]
### Feature request

**Is your feature request related to a problem? Please describe.**

Greedy assisted decoding in Transformers uses strict argmax verification: the entire draft
suffix is discarded at the first position where the draft token differs from the target
argmax. This is lossless, but it rejects near-optimal draft tokens whose logit gap to the
argmax is negligible, capping the practical speedup of speculative decoding for greedy users.

PR #45979 (`assistant_ensemble_weight`) added a *lossy* verification option for the
sampling path by mixing the target and draft distributions. Greedy decoding, however, still
lacks a deterministic, explicitly-bounded relaxation mechanism.

**Describe the solution you'd like**

Add **Approximate Speculative Decoding (ASD)** 鈥?a training-free, verifier-side acceptance
policy for greedy assisted decoding. For a draft token `x_i` and target logits `z_i`, define
the local regret `r_i = max_v z_i(v) - z_i(x_i)` (`r_i = 0` iff `x_i` is the target argmax).
A draft prefix is accepted only while ALL of the following hold:

1. Request-level cumulative regret stays within a budget `B`;
2. Local regret satisfies `r_i / (K - i) <= g`, where `K` is the candidate length
   (suffix-value weighting: later draft positions get less slack);
3. The number of relaxed mismatches in the current block does not exceed `m`.

Verification stops at the first infeasible position and commits the target argmax bonus
token as usual. The policy is fully deterministic (no RNG) and **exactly reduces to strict
greedy verification when `B = 0` or `m = 0`** (all `r_i` must be 0, i.e. strict argmax
prefix matching) 鈥?our gold-standard correctness check.

**Proposed API** (three scalar parameters, mirroring the `assistant_ensemble_weight` style;
naming open to maintainer feedback):

```python
outputs = model.generate(
    **inputs,
    assistant_model=assistant_model,
    assistant_asd_budget=2.0,          # float >= 0, request-level cumulative regret budget; None = disabled (lossless)
    assistant_asd_local_ratio=0.25,    # float >= 0, per-token regret / suffix-value cap
    assistant_asd_max_mismatches=2,    # int >= 0, max relaxed mismatches per block
)
```

**How is this different from `assistant_ensemble_weight` (#45979)?**

| Aspect | `assistant_ensemble_weight` | ASD |
|---|---|---|
| Mechanism | verifies against a static mixture `w*p + (1-w)*q` | deterministic per-token regret rule, distributions untouched |
| Error control | global TV-bias bound `w*TV(q,p)` | explicit per-request budget + suffix-value weighting + per-block mismatch cap (triple control) |
| Target path | mainly `do_sample=True` | greedy decoding (no RNG, reproducible) |
| Lossless fallback | only `None` | `B=0` or `m=0` *is* strict verification by construction |
| Draft logits | required | not needed (regret uses target logits only) |

The two are orthogonal and could coexist.

**Empirical results** (research evaluator, pure PyTorch, Qwen3-14B + DSpark block-7 draft,
8xL20 GPUs, greedy):

| Setting | Throughput | GSM8K |
|---|---|---|
| Strict verification | 66.21 tok/s | 79.682% |
| ASD (B=2.0625) | 75.04 tok/s (+13.34%) | 79.454% (-0.227pp) |
| ASD (B=0) | token-identical to strict on all 1319 requests | identical |

**Implementation scope** (~150 lines of library code, no external dependency):

1. Three scalar parameters in `GenerationConfig` + checks in `validate()`;
2. A small acceptance helper next to `_speculative_sampling` in
   `src/transformers/generation/utils.py`, called from the greedy branch (Case 2) of
   `GenerationMixin._assisted_decoding()`; a request-level budget accumulator in the decode
   loop; a `ValueError` guard for prompt lookup decoding (no candidate logits) in
   `_get_candidate_generator`;
3. Unit tests added to the existing `tests/generation/test_utils.py` and
   `tests/generation/test_configuration_utils.py` (no new test files), plus a short section
   in `docs/source/en/assisted_decoding.md`.

**References**

- Paper: ASD: Approximate Speculative Decoding, https://arxiv.org/abs/2608.03447
- Reference implementation (Apache-2.0): https://github.com/Kissmetothemoon/ASD

I am one of the paper authors and happy to implement this and submit a PR if there is
interest from the maintainers.
